### DISCUSSION TITLE
Versioning npm modules if no commits to main in CI?

### DISCUSSION THREAD

**Discussion by mpsanchis** [nrwl/nx#17186] [General]
Hi everyone!

My team is running into an issue, and we thought we’d give it a try to ask here.

## Preamble
Our git workflow prevents commits to main without merge requests, so we are avoiding creating extra commits in CI. This means that all our npm modules have `“version”: “0.0.0”` in their `package.json`. We use `@jscutlery/semver` to calculate versions, but they’re never committed. They are calculated in CI, and the correct version is published in the npm registry.

## Problem
We have project A that depends on project B. Whenever B is changed, it triggers the version calculation in A, and bumps a patch version if nothing changed in A. However, when A is changed and built, it builds a package in `dist/` that depends on B with version 0.0.0, which is not available in the registry.
This means that our published package A has a faulty dependency.

## Possible solution
We’re thinking of running the versioning target two times:
1. for the modules from `nx affected`: would bump patch versions for modules with no changes (like A when only B has been modified)
2. for the “unaffected” nx modules: would re-calculate versions of modules, without bumping (like getting B’s current version when A has been modified)

## Questions
- Do you know of a better way to approach this problem, if we want to continue avoiding commits in CI that modify the `package.json`?
- Is there a clean way of getting the unaffected nx projects? I can get the values in `nx print-affected` (would be `projectGraph.nodes` excluding `projects`), but maybe there is an alternative.

Thank you!

**Comment by kevindice** [comment]
Hi @mpsanchis - would you mind sharing the args you are passing to `@jscutlery/semver` for context?

Wanted to make sure you were aware of the `--trackDeps` arg along with the accompanying `nx.json` config suggestion - both here:  https://github.com/jscutlery/semver#tracking-dependencies

Since `@jscutlery/semver` works off of git tags, am I correct in thinking that you still push up the git tags to GitHub even though the commits themselves don't end up in `main`/`master`?

You are likely fine continuing not to push these commits to `main`, but if the only reason preventing this is the branch protection rules, you can get around them by granting a non-human user the `bypass branch protections` permission and configure Circle with this user.

One issue though is that `@jscutlery/semver` creates one commit for each module in independent mode, and this can get a bit ridiculous with several dozen modules.  I've decided to squash these commits and move all the tags to the single squashed commit like so before pushing to `main`:

```bash
#!/bin/bash

# Configure the git user who will be making the squashed commit
git config user.name $(git log -1 --pretty=format:'%an')
git config user.email $(git log -1 --pretty=format:'%ae')

# Get the commit hash before we run the release stuff.
HEAD_BEFORE_RELEASE=$(git rev-parse HEAD)
echo "HEAD_BEFORE_RELEASE" "$HEAD_BEFORE_RELEASE"

# For each affected package, bump the release version.
# This results in one commit FOR EACH affected package,
# which is unfortunate.  So, below, we will merge them
# and then relocate all of the tags.
yarn nx affected --base=last-release --target=version --parallel=1

# Get the commit hash after the release, before squashing
HEAD_BEFORE_SQUASH=$(git rev-parse HEAD)
echo "HEAD_BEFORE_SQUASH" "$HEAD_BEFORE_SQUASH"

# Get the number of commits generated by the nx release plugin script
COUNT=$(git rev-list --count $HEAD_BEFORE_RELEASE..$HEAD_BEFORE_SQUASH)

# Exit here if there were no packages released.
if [ "$COUNT" -eq "0" ]; then
  echo "No packages to release this time.  Skipping...";
  exit 0;
fi

# Squash all the release commits into a single one.
git reset --soft $HEAD_BEFORE_RELEASE
git add .
git commit -m "chore: release"

# But now, we still need to move all of the tags.  So, we will
# iterate through them.

# Due to https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#createevent
# (GitHub does not call the webhook when you push more than 3 tags at once)
# ...we must batch them.

# Build an array of tags which come from the set of new commits that we just squashed.
# Also delete the tags since we want to avoid pushing more than 3 at once.
declare -a NEW_TAGS
for commit in $(git --no-pager rev-list $HEAD_BEFORE_RELEASE..$HEAD_BEFORE_SQUASH)
do
  TAG=$(git tag --points-at $commit)
  echo "Tag(s)" "$TAG"
  NEW_TAGS+=("$TAG")
  git tag -d $TAG
done

# Push up our squashed commit.
# If the push is rejected due to a race condition (two PRs merged too close together)
# then we just bail now and do not update any tags.  The next release commit will
# encompass all changes and changelog entries, resulting in the appropriate modules
# eventually getting the tags they are due.
#
# We can go fast, and this is our seat belt. 🤘
git push || exit 0;

# Move the previous-release tag and push it
git checkout last-release
git tag -f previous-release
git push -f origin previous-release

# Move the last-release tag and push it
git checkout main
git tag -f last-release
git push -f origin last-release

# Now, we re-create the package-specific tags for this release
# and push them in batches of at most 3.
echo "Re-creating tags and pushing in batches of 3!"
TAG_BATCH_COUNTER=0
for new_tag in "${NEW_TAGS[@]}"
do
  echo $new_tag
  git tag $new_tag
  TAG_BATCH_COUNTER=$((TAG_BATCH_COUNTER+1))
  if [ "$TAG_BATCH_COUNTER" -eq "3" ]; then
    git push origin --tags
    TAG_BATCH_COUNTER=0
  fi
done

# Send the last batch of less than 3 tags, if there is one.
git push origin --tags
```

Note:  I have to push the new tags in batches of at most 3 because I fire off CircleCi release jobs based on some of these tags, and there is a limitation with GitHub where webhooks do not get called when more than 3 tags are pushed at once.

To address your 2nd question:  Since you don't push these commits currently, each package.json file is not kept up to date with the current version number.  So, the only place that this version state resides would be in git tags.  To access it, you would have to do what `@jscutlery/semver` does which is run `git --no-pager tag` and filter through all the tags, finding the highest version number for each package, and use that to set the appropriate dependency values.  That seems tedious and potentially redundant to stuff done internally by the semver plugin.  It may be easier to push the generated commits in the first place.

I'm curious though whether the `--trackDeps` argument and the following config
```
{
  "targetDependencies": {
    "version": [
      {
        "target": "version",
        "projects": "dependencies"
      }
    ]
  }
}
```
solves your issue by essentially running the version target on the extremities of the dependency graph first before moving on to dependents.  Since the dependencies are version stamped first, I would imagine maybe this solves your problem because then "project B" would have a real version number and not `0.0.0` when "project A" has the version target run on it.

**Reply by kevindice** [reply]
I was under the impression that `@jscutlery/semver` requires the package version git tags to lie in the history of the `main` branch - so I wouldn't have expected your current approach of not pushing these commits to work.  Does it?

**Comment by mpsanchis** [comment]
Hi @kevindice:

First of all, a big thank you for such a detailed response. Very much appreciate it.

I have been working most of the day on this, and I think I’m getting somewhere. I’ll reply in two parts:

## Reply to questions

So, more about our setup:

> would you mind sharing the args you are passing to @jscutlery/semver for context?
> am I correct in thinking that you still push up the git tags to GitHub even though the commits themselves don't end up in main/master?

Yes, you are correct. Currently, we are using these flags (mainly):
```
        "tagPrefix": "{projectName}@",
        "skipCommit": true,
        "allowEmptyRelease": true,
        "postTargets": ["{projectName}:prepare-envvars"]
```

Unfortunately, bypassing branch protection is not an option for us, due to rules that I can’t change :( . However, I think we will be able to find a solution.


> Wanted to make sure you were aware of the --trackDeps arg

To be honest I hadn’t tried it, but we are using `nx affected` together with `allowEmptyRelease`. Therefore, if B is changed, then A’s `version` target is also triggered, and it bumps the version with a patch increase. I tried with `trackDeps` and didn’t see any difference for our use case.

> I wouldn't have expected your current approach of not pushing these commits to work. Does it?

As you correctly said, `@jscutlery/semver` requires git tags to lie in the history of the main branch, and we still push up the tags to the remote repo, so that’s why it’s able to calculate the version correctly.

## Next Steps

As I said, we’re still going to avoid committing in CI. However, what we were not doing correctly was setting the `targetDependencies` (`targetDefaults` from nx 16). We are currently trying the following, when it comes to building Typescript modules that end up in an npm registry:
1. Always call the `version` target before the `build` (build dependsOn version)
2. Always call the `build` target of your dependencies before calculating your version (version dependsOn ^build)

This strategy will hopefully always modify the `package.json` locally in the machine that runs the job in CI before the code is packaged into `/dist` ready to be pushed to the registry.

Again, thank you for the hints. They helped a lot, since I started trying out things that I didn’t think of before.
